src/op.rs

```diff
@@ -0,0 +1,1288 @@
+            callbacks: callbacks,
+            max_iterations: None,
+            phantom: PhantomData,
+        }
+    }
+
+    /// Set an initial value to be pushed on the DWARF expression
+    /// evaluator's stack.  This can be used in cases like
+    /// `DW_AT_vtable_elem_location`, which require a value on the
+    /// stack before evaluation commences.
+    pub fn set_initial_value(&mut self, value: u64) {
+        self.initial_value = Some(value);
+    }
+
+    /// Set the enclosing object's address, as used by
+    /// `DW_OP_push_object_address`
+    pub fn set_object_address(&mut self, value: u64) {
+        self.object_address = Some(value);
+    }
+
+    /// Set the maximum number of iterations to be allowed by the
+    /// expression evaluator.  The default is `None`.  This value can
+    /// be set to avoid denial of service attacks by bad DWARF
+    /// bytecode.
+    pub fn set_max_iterations(&mut self, value: u32) {
+        self.max_iterations = Some(value);
+    }
+
+    /// Evaluate a DWARF expression.
+    pub fn evaluate(&mut self) -> ParseResult<Vec<Piece<'input>>>
+        where Endian: Endianity
+    {
+        // Stack operations are done on word-sized values.  We do all
+        // operations on 64-bit values, and then mask the results
+        // appropriately when popping.
+        let addr_mask = if self.address_size == 8 {
+            !0u64
+        } else {
+            (1 << 8 * self.address_size as u64) - 1
+        };
+
+        let mut stack = Vec::<u64>::new();
+        if let Some(value) = self.initial_value {
+            stack.push(value);
+        }
+
+        let pop = |stack: &mut Vec<u64>| {
+            match stack.pop() {
+                Some(value) => Ok(value & addr_mask),
+                None => Err(Error::NotEnoughStackItems),
+            }
+        };
+
+        let mut iteration = 0;
+
+        // Stack of expressions we're evaluating.  We push a new entry on
+        // this stack each time we call a subroutine.
+        let mut expression_stack = Vec::new();
+
+        // The results.
+        let mut result = Vec::new();
+
+        let mut bytecode = self.bytecode;
+        let mut pc = EndianBuf::<Endian>::new(self.bytecode);
+
+        'eval: loop {
+            while pc.len() == 0 {
+                match expression_stack.pop() {
+                    Some((newpc, newbytes)) => {
+                        pc = newpc;
+                        bytecode = newbytes;
+                    }
+                    None => break 'eval,
+                }
+            }
+
+            iteration += 1;
+            if let Some(max_iterations) = self.max_iterations {
+                if iteration > max_iterations {
+                    return Err(Error::TooManyIterations);
+                }
+            }
+
+            let mut terminated = false;
+            let mut piece_end = false;
+            let (newpc, operation) = try!(decode(pc, bytecode, self.address_size, self.format));
```

Comment:
So in a loop or recursive function or any code that ends up getting executed more than once, we end up decoding the same bytecodes repeatedly?

I guess it is possible to jump between bytecodes and operands and have that get interpreted as valid expressions as well?

I think it would be better if we only decode the operations once and then just jump between decoded operations with a map from byte offset -> decoded opcode index.

Even better would be [separating syntactic analysis from execution](https://mitpress.mit.edu/sicp/full-text/sicp/book/node83.html) by "compiling" operations into lambdas so we skip the "which opcode are we interpreting right now?" check on each iteration (jimb calls this "lambda chaining", but I haven't heard that term anywhere else). Maybe rust's discriminated unions already give us most of the benefit of this, compared to scheme where we would need to repeatedly grab + check operands...

I think we can safely do this stuff as follow ups (unless you think the former is easy enough to land with this initial PR, the latter is far in follow up territory IMO).